File: scripts/publisher.py
import os
import requests
from notion_client import Client

# --- 1. CONFIGURATION (Gets variables from GitHub Secrets) ---
NOTION_TOKEN = os.environ.get("NOTION_TOKEN")
NOTION_DATABASE_ID = os.environ.get("NOTION_DB_ID")
LINKEDIN_ACCESS_TOKEN = os.environ.get("LINKEDIN_ACCESS_TOKEN")
LINKEDIN_PERSON_URN = os.environ.get("LINKEDIN_PERSON_URN") 

# Initialize Notion Client
if not NOTION_TOKEN:
    raise ValueError("NOTION_TOKEN environment variable not set.")
notion = Client(auth=NOTION_TOKEN)

# --- 2. NOTION DATA RETRIEVAL ---
def get_ready_to_post_article():
    """
    Queries the Notion database for the first page with Status == 'Ready to Post'.
    Extracts the Title, LinkedIn Copy, and Public URL properties.
    """
    print("--- Checking Notion Database for Ready-to-Post Articles ---")
    try:
        response = notion.data_sources.query(
            data_source_id=NOTION_DATABASE_ID,
            filter={
                "property": "Status",
                "select": {
                    "equals": "Ready to Post"
                }
            },
            sorts=[
                {
                    "property": "Name",
                    "direction": "ascending"
                }
            ]
        )
    except Exception as e:
        print(f"Error querying Notion database: {e}")
        return None

    if not response.get('results'):
        print("SUCCESS: No articles ready to post. Exiting script.")
        return None

    page = response['results'][0]
    properties = page['properties']
    page_id = page['id']
    
    # --- Property Extraction Logic ---
    
    # 1. Extract the Title (Name) property - Type: 'title'
    article_title = properties.get('Name', {}).get('title', [{}])[0].get('text', {}).get('content', '')
    
    # 2. Extract the LinkedIn Copy property - Type: 'rich_text'
    linkedin_copy = properties.get('LinkedIn Copy', {}).get('rich_text', [{}])[0].get('text', {}).get('content', '')
    
    # 3. Extract the Public URL property - Type: 'url'
    public_url = properties.get('Public URL', {}).get('url', '')

    if not all([article_title, linkedin_copy, public_url]):
        print(f"ERROR: Missing required fields on Notion page ID: {page_id}. Title, Copy, or URL is empty.")
        return None
        
    print(f"Found article to post: '{article_title}'")
    
    return {
        'page_id': page_id,
        'title': article_title,
        'copy': linkedin_copy,
        'url': public_url
    }

# --- 3. LINKEDIN POSTING (RICH LINK SHARE UPDATE) ---
def post_rich_link_to_linkedin(article_data):
    """Posts an engaging link share update to LinkedIn."""
    print("--- Attempting to Post to LinkedIn ---")
    
    api_url = "https://api.linkedin.com/v2/ugcPosts"
    headers = {
        "Authorization": f"Bearer {LINKEDIN_ACCESS_TOKEN}",
        "Content-Type": "application/json",
        "X-Restli-Protocol": "2.0.0"
    }

    payload = {
        "author": LINKEDIN_PERSON_URN,
        "lifecycleState": "PUBLISHED",
        "specificContent": {
            "com.linkedin.ugc.ShareContent": {
                "shareCommentary": {
                    "text": article_data['copy']
                },
                "shareMediaCategory": "ARTICLE",
                "media": [
                    {
                        "status": "READY",
                        "originalUrl": article_data['url'], 
                        "title": {
                            "text": article_data['title'] 
                        }
                    }
                ]
            }
        },
        "visibility": {
            "com.linkedin.ugc.MemberNetworkVisibility": "PUBLIC"
        }
    }

    try:
        response = requests.post(api_url, headers=headers, json=payload)
        response.raise_for_status() # Raise an exception for bad status codes (4xx or 5xx)

        print(f"SUCCESS: Successfully posted to LinkedIn! Status Code: {response.status_code}")
        return True
    
    except requests.exceptions.HTTPError as err:
        print(f"ERROR: LinkedIn post failed with HTTP error: {err}")
        print(f"Response Body: {response.text}")
        return False
    except Exception as e:
        print(f"An unexpected error occurred during posting: {e}")
        return False

# --- 4. UPDATE NOTION STATUS ---
def update_notion_status(page_id, new_status):
    """Updates the Status property of the Notion page."""
    print(f"--- Updating Notion Page Status to '{new_status}' ---")
    
    try:
        notion.pages.update(
            page_id=page_id,
            properties={
                "Status": {
                    "select": {
                        "name": new_status
                    }
                }
            }
        )
        print("SUCCESS: Notion status updated.")
        return True
    except Exception as e:
        print(f"ERROR: Could not update Notion status for page {page_id}: {e}")
        return False


# --- MAIN EXECUTION ---
if __name__ == "__main__":
    article = get_ready_to_post_article()
    
    if article:
        if post_rich_link_to_linkedin(article):
            # If successful, mark as Published
            update_notion_status(article['page_id'], "Published")
        else:
            # If failed, mark as Post Failed
            update_notion_status(article['page_id'], "Post Failed")

File: publisher.js
require('dotenv').config();
const { Client } = require('@notionhq/client');
const axios = require('axios');

// Configuration from Environment Variables
const notion = new Client({ auth: process.env.NOTION_TOKEN });
const databaseId = process.env.NOTION_DB_ID;
const linkedinAccessToken = process.env.LINKEDIN_ACCESS_TOKEN;
const linkedinUrn = process.env.LINKEDIN_PERSON_URN;

async function processLinkedinPosts() {
  try {
    // 1. Query Notion for pages with Status "Ready to Post"
    const response = await notion.databases.query({
      database_id: databaseId,
      filter: {
        property: "Status",
        select: {
          equals: "Ready to Post",
        },
      },
    });

    for (const page of response.results) {
      const pageId = page.id;
      const title = page.properties['Name']?.title[0]?.plain_text || "Untitled";
      const linkedinCopy = page.properties['LinkedIn Copy']?.rich_text.map(part => part.plain_text).join('');
      const publicUrl = page.properties['Public URL']?.url;
      
      const imagesProperty = page.properties['Images']?.files || [];
      const imageFile = imagesProperty.length > 0 ? imagesProperty[0] : null;
      const imageUrl = imageFile ? (imageFile.file?.url || imageFile.external?.url) : null;

      try {
        if (!linkedinCopy) {
          throw new Error("Missing LinkedIn Copy");
        }

        console.log(`Attempting to post: ${title}...`);
        const isSuccess = await postToLinkedin(linkedinCopy, publicUrl, imageUrl);

        if (isSuccess) {
          // SUCCESS: Move to Published
          await notion.pages.update({
            page_id: pageId,
            properties: {
              'Status': { select: { name: 'Published' } }
            },
          });
          console.log(`✅ Successfully published: ${title}`);
        } else {
          throw new Error("LinkedIn API rejected the post");
        }
      } catch (error) {
        console.error(`❌ Failed to post "${title}":`, error.message);

        // FAILURE: Move to Post Failed
        await notion.pages.update({
          page_id: pageId,
          properties: {
            'Status': { select: { name: 'Post Failed' } }
          },
        });
      }
    }
  } catch (error) {
    console.error("Error querying Notion database:", error.body || error.message);
  }
}

async function uploadImageToLinkedin(imageUrl) {
  try {
    // 1. Register the upload
    const registerResponse = await axios.post(
      'https://api.linkedin.com/v2/assets?action=registerUpload',
      {
        registerUploadRequest: {
          recipes: ['urn:li:digitalmediaRecipe:feedshare-image'],
          owner: linkedinUrn,
          serviceRelationships: [
            {
              relationshipType: 'OWNER',
              identifier: 'urn:li:userGeneratedContent',
            },
          ],
        },
      },
      {
        headers: {
          'Authorization': `Bearer ${linkedinAccessToken}`,
          'X-Restli-Protocol-Version': '2.0.0',
        },
      }
    );

    const uploadUrl = registerResponse.data.value.uploadMechanism['com.linkedin.digitalmedia.uploading.MediaUploadHttpRequest'].uploadUrl;
    const assetId = registerResponse.data.value.asset;

    // 2. Download the image from Notion
    const imageBuffer = await axios.get(imageUrl, { responseType: 'arraybuffer' });

    // 3. Upload the binary data to LinkedIn
    await axios.put(uploadUrl, imageBuffer.data, {
      headers: {
        'Authorization': `Bearer ${linkedinAccessToken}`,
        'Content-Type': 'image/jpeg',
      },
    });

    return assetId;
  } catch (error) {
    console.error("❌ Image Upload Error:", error.response?.data || error.message);
    return null;
  }
}

async function postToLinkedin(text, url, imageUrl) {
  try {
    let shareMediaCategory = 'NONE';
    let media = [];
    let finalText = text;

    if (imageUrl) {
      const assetId = await uploadImageToLinkedin(imageUrl);
      if (assetId) {
        shareMediaCategory = 'IMAGE';
        media = [{
          status: 'READY',
          media: assetId,
        }];
        // If there's also a link, append it to the text as LinkedIn's ugcPosts 
        // doesn't support both Image and Article media in one share.
        if (url) {
          finalText += `\n\n${url}`;
        }
      }
    }

    if (shareMediaCategory === 'NONE' && url) {
      shareMediaCategory = 'ARTICLE';
      media = [{
        status: 'READY',
        originalUrl: url,
      }];
    }

    const response = await axios.post(
      'https://api.linkedin.com/v2/ugcPosts',
      {
        author: linkedinUrn,
        lifecycleState: 'PUBLISHED',
        specificContent: {
          'com.linkedin.ugc.ShareContent': {
            shareCommentary: { text: finalText },
            shareMediaCategory: shareMediaCategory,
            media: media,
          },
        },
        visibility: { 'com.linkedin.ugc.MemberNetworkVisibility': 'PUBLIC' },
      },
      {
        headers: {
          'Authorization': `Bearer ${linkedinAccessToken}`,
          'X-Restli-Protocol-Version': '2.0.0',
          'Content-Type': 'application/json'
        },
      }
    );
    console.log("LinkedIn Response Status:", response.status);
    return response.status === 201;
  } catch (err) {
    // THIS PART IS KEY: It prints the specific error from LinkedIn's server
    if (err.response) {
      console.error("❌ LinkedIn API Error (Data):", JSON.stringify(err.response.data, null, 2));
      console.error("❌ LinkedIn API Error (Status):", err.response.status);
    } else if (err.request) {
      console.error("❌ LinkedIn API Error (No Response): No response received from LinkedIn. Check your internet or URN format.");
    } else {
      console.error("❌ Error Message:", err.message);
    }
    return false;
  }
}

processLinkedinPosts();